backend/src/qs/key_package.rs

```diff
@@ -96,17 +96,20 @@ mod persistence {
             let mut transaction = connection.begin().await?;
 
             let encrypted_key_package_option = sqlx::query_scalar!(
-                r#"WITH deleted_package AS (
-                    DELETE FROM key_packages
-                    USING qs_client_records qcr
+                r#"WITH to_delete AS (
+                    SELECT id FROM key_packages
+                    INNER JOIN qs_client_records qcr
+                        ON qcr.client_id = key_packages.client_id
                     WHERE
-                        key_packages.client_id = qcr.client_id
-                        AND key_packages.client_id = $1
+                        key_packages.client_id = $1
                         AND qcr.user_id = $2
-                    RETURNING key_packages.id, key_packages.encrypted_key_package
+                    LIMIT 1
+                    FOR UPDATE SKIP LOCKED
                 )
-                SELECT encrypted_key_package as "eap: _" FROM deleted_package
-                FOR UPDATE SKIP LOCKED"#,
+                DELETE FROM key_packages
+                WHERE id IN (SELECT id FROM to_delete)
+                RETURNING encrypted_key_package AS "eap: _"
+                "#,
```

Comment:
I'm not sure the query was better (or even correct) before, but don't we now have the problem that two concurrent queries could select the same KeyPackage in the first query? In the second query one would then fail when trying to delete it. Ideally, we would already lock in the first query, which is what I think I was trying to do with the original one.

Reply:
FOR UPDATE SKIP LOCKED is now part of the `to_delete` select. I am not an expert in postgres, but as far as I understand, it will be locked as long the transaction exists.

NB: A single query is always transactional, so there is no need for a transaction here, which a little bit simplifies the code.

Reply:
Ah, you're right of course! My bad.